docs/adrs/003-Retry_endpoint_design.md

```diff
@@ -0,0 +1,162 @@
+* Non blocking job submission / (Possibly htcondor submit)
+* Add failure conditions in run method
+* Add thread to perform actions based on toggle
+
+
+### Questions
+
+#### Answered:
+#### Q: should the number of retries of a job be limited, and if so, where? e.g. a max_retries field in the parent job? wait and see whether people attempt to rerun jobs that have already failed nine zillion times?
+A: Make a ticket for this and add to backlog
+
+#### Q: How do we prevent jobs with identical parameters from being rerun more than once within a retry_jobs request?
+A: We have decided to allow multiple jobs with the same params to be re-run in the same `retry_jobs` request.
+
+#### Q: How do we find the most recent retry of a job?
+A: The client using the ee2 API would have to figure it out using the `retry_parent` and job creation date fields. (Unless we added other fields to help with this)
+
+#### Q: How do we ensure that the app version is correctly run each time when submitting from the narrative?
+A: We would need to change the narrative to submit the git commit hash instead of a version tag
+
+#### Q: How do we handle DB consistency during retry failure? 
+Looks like the options are
+* implement db integrity checks and two-phase commits for making the relationships between a job, its `retry_parent`, and the batch container
+* accept that the db info may be incomplete and write workarounds into the clients
+* (upgrade to Mongo 4.4 for better transaction support)
+
+
+##### Q: Do we want to support ResourceRequirements
+A: Probably not in the short term
+
+
+#### Q: how to prevent incorrect parent-child relationships being created -- should the client be allowed to specify a parent ID? Is it currently possible to add a new child to a parent job if the child is a new job, rather than an existing job ID / set of params that is being rerun?
+A: Not necessarily relevant to this endpoint, more of a `run_job_batch` endpoint question. Currently the `retry_parent` and `parent_job_id` are looked up from the ee2 record on retry, and not specified in this endpoint.
+
+#### Shorter Q and A
+
+    Should we track a retry count? (Done)
+    Should users see this retry count? (Unknown TBD)
+    Are retried jobs saved in some sort of data structure linking them, possibly indirectly, to the parent job or are they orphaned? (Yes, retry_parent)
+    If the former, is the retry relationship linear or a tree? E.g. what happens if there are two simultaneous calls to retry a job? (Tree, simultaneous jobs run)
+    Should it be at least theoretically possible to see the list of retried jobs in order? (It is possible by sorting on creation date)
+    Should there be a maximum retry count? Or a warning that more retries are not likely to help?  (Unknown TBD)
+    Can a job in states other than failed or canceled be retried? Or should the user be required to cancel a job before it can be retried? (Job must be in Error/Cancel state)
```

Comment:
Do you want to turn these into proper Qs and As, or alternatively incorporate the information into the specs above somewhere?

Reply:
I don't remember the reason anymore about why there were shorter questions anymore and why they didn't make it into the big list. Maybe it's because they were already answered? 
I think it's fine the way it is, but can change it if you want.

Reply:
@ialarmedalien Do you require any changes here?

Reply:
Nope.